### ISSUE TITLE
Limit the CommandLineInstruction to max possible length

### ISSUE THREAD

**Issue by steffen-wilke** [BrightLight/RepoCop#19]
There are limitations on the length of the command that `Process.Start(...)` produces.
If such a limit is exceeded e.g. there might be a lot of `affectedpaths` in a commit and the instruction tries to pass these as an #affectedpaths# argument to an external progam.

From https://stackoverflow.com/questions/1200789/c-sharp-process-start-parameters-truncated
> On computers running Microsoft Windows XP or later, the maximum length of the string that you can use at the command prompt is 8191 characters...

Currently, this results in an exception:
```
29.01.2019-14:45:16,185 [1] ERROR Silverseed.RepoCop.Subversion.Program.Main(C:\projects\repocop\source\Console\Program.cs:133) - An error occured
System.ComponentModel.Win32Exception (0x80004005): Der Dateiname oder die Erweiterung ist zu lang
   bei System.Diagnostics.Process.StartWithShellExecuteEx(ProcessStartInfo startInfo)
   bei System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   bei Silverseed.RepoCop.CommandLineInstruction.ExecuteFireAndForgetInstruction() in C:\projects\repocop\source\CommandLineInstruction.cs:Zeile 100.
   bei Silverseed.RepoCop.MacroInstruction.<>c__DisplayClass2_0.<InternalExecute>b__0(Instruction x) in C:\projects\repocop\source\MacroInstruction.cs:Zeile 35.
   bei System.Collections.Generic.List`1.ForEach(Action`1 action)
   bei Silverseed.RepoCop.MacroInstruction.InternalExecute() in C:\projects\repocop\source\MacroInstruction.cs:Zeile 36.
   bei Silverseed.RepoCop.MacroInstruction.<>c__DisplayClass2_0.<InternalExecute>b__0(Instruction x) in C:\projects\repocop\source\MacroInstruction.cs:Zeile 35.
   bei System.Collections.Generic.List`1.ForEach(Action`1 action)
   bei Silverseed.RepoCop.MacroInstruction.InternalExecute() in C:\projects\repocop\source\MacroInstruction.cs:Zeile 36.
   bei Silverseed.RepoCop.Subversion.Program.Main(String[] args) in C:\projects\repocop\source\Console\Program.cs:Zeile 133.
```

**Possible  solutions I see here:**

1. Truncate the arguments string at `8191 - (filename + 1)` (which might result in undefined behavior of consuming applications.
2. Dump the arguments to a temporary file and pass a reference to that file as argument (e.g. `myapplication.exe @arguments-rev123.txt` (Notice the **@** to identify a file reference)

**Comment by BrightLight** [comment]
Not trivial because of details :(
For instance, pass all arguments as a file or should the file only contain the list of affected paths? If all arguments, then a format needs to be defined (maybe XML?) and of course the calling app needs to be able to parse it. Therefore the format should probably be fairly simple. A simply approach might be to still pass almost all arguments in the command line, except for the affected paths which are passed by file. The file format could be very simply: one line for each affected path.
This this doesn't work if the svn action type (added, deleted, etc.) is to be included.

I'm not a fan of just truncating the command line as this might cause an invalid call. If anything, I would try to truncate the part with the affected paths. But that requires quite a rewrite in how the command line is built I think.

**Comment by BrightLight** [comment]
I've now added several "@" tokens:

 * #@affectedfiles#
 * #@affectedfilesdetailed#
 * #@affectedpaths#
 * #@affectedpathsdetailed#

If "@" is used, the content of the token is written into a file and the token is replaced with the name (and path) of this file.

This is esp. useful for `CommandLineInstruction` and the "affectedfiles" token family. If a CommandLineInstruction used one of the "affectedfiles" tokens, all affected files were written into the command line. If a lot of files were affected by the commit, this could lead to a command line that is larger than what Windows supports and an error would occur.
Now we support the following tokens:

The "@" indicates that the content of the token will not be passed directly in the command line. Instead, the content is written to a file (in the system temp folder) and the path and filename of that file are passed in the command line (with a leading "@" to indicate that this is a file reference)
The filename is built using the name of the token and the revision number. Example: "affectedfilesdetailed-12345"

[Documentation](https://github.com/BrightLight/RepoCop/wiki) was updated accordingly.
